File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Pitch/PitchPointExplorerDemo.swift
import SwiftUI
import VoxaTrace

// MARK: - Demo 4: PitchPoint Explorer

/// Showcases all PitchPoint computed properties in real-time.
struct PitchPointExplorerDemo: View {
    @State private var selectedAlgorithm: Int = 0 // YIN default
    @State private var detector: CalibraPitch.Detector?
    @State private var recorder: SonixRecorder?
    @State private var isRecording = false
    @State private var currentPitch: PitchPoint?
    @State private var amplitude: Float = 0.0
    @State private var modelLoaderConfigured = false

    private let algorithms: [(algorithm: PitchAlgorithm, name: String)] = [
        (.yin, "YIN"),
        (.swiftF0, "SwiftF0")
    ]

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("PitchPoint Properties")
                .font(.subheadline)
                .fontWeight(.semibold)

            Text("See all computed properties of PitchPoint in real-time.")
                .font(.caption)
                .foregroundColor(.secondary)

            // Algorithm picker
            HStack {
                Text("Algorithm:")
                    .font(.caption)
                Picker("Algorithm", selection: $selectedAlgorithm) {
                    ForEach(0..<algorithms.count, id: \.self) { index in
                        Text(algorithms[index].name).tag(index)
                    }
                }
                .pickerStyle(.segmented)
                .onChange(of: selectedAlgorithm) { _ in
                    if isRecording {
                        stopRecording()
                    }
                    detector?.release()
                    detector = nil
                }
            }

            // Main display
            pitchDisplaySection

            Divider()

            // Properties inspector
            propertiesInspector

            Divider()

            // Control
            HStack {
                Button(isRecording ? "Stop" : "Start") {
                    if isRecording {
                        stopRecording()
                    } else {
                        startRecording()
                    }
                }
                .buttonStyle(.borderedProminent)
            }

            Divider()

            apiInfoSection
        }
        .onDisappear {
            stopRecording()
            cleanup()
        }
    }

    private var pitchDisplaySection: some View {
        VStack(spacing: 16) {
            // Large note display
            VStack {
                if let pitch = currentPitch, pitch.isSinging, let note = pitch.note {
                    Text(note)
                        .font(.system(size: 60, weight: .bold, design: .rounded))
                        .foregroundColor(.blue)

                    Text(String(format: "%.1f Hz", pitch.pitch))
                        .font(.title3)
                        .foregroundColor(.secondary)
                } else {
                    Text("-")
                        .font(.system(size: 60, weight: .bold, design: .rounded))
                        .foregroundColor(.gray)

                    Text("Not singing")
                        .font(.title3)
                        .foregroundColor(.secondary)
                }
            }
            .frame(maxWidth: .infinity)
            .padding()
            .background(Color(.secondarySystemBackground))
            .cornerRadius(12)

            // Tuning indicator
            if let pitch = currentPitch, pitch.isSinging {
                TuningIndicatorView(centsOff: pitch.centsOff, tuning: pitch.tuning)
            }
        }
    }

    private var propertiesInspector: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Properties Inspector")
                .font(.subheadline)
                .fontWeight(.semibold)

            LazyVGrid(columns: [
                GridItem(.flexible()),
                GridItem(.flexible())
            ], spacing: 8) {
                PropertyRow(name: "pitch", value: currentPitch.map { String(format: "%.1f Hz", $0.pitch) } ?? "-")
                PropertyRow(name: "isSinging", value: currentPitch.map { $0.isSinging ? "true" : "false" } ?? "-")
                PropertyRow(name: "note", value: currentPitch?.note ?? "-")
                PropertyRow(name: "midiNote", value: currentPitch.map { $0.isSinging ? "\($0.midiNote)" : "-" } ?? "-")
                PropertyRow(name: "centsOff", value: currentPitch.map { $0.isSinging ? "\($0.centsOff)" : "-" } ?? "-")
                PropertyRow(name: "tuning", value: currentPitch.map { tuningString($0.tuning) } ?? "-")
                PropertyRow(name: "reliability", value: currentPitch.map { String(format: "%.2f", $0.reliability) } ?? "-")
                PropertyRow(name: "confidence", value: currentPitch.map { String(format: "%.2f", $0.confidence) } ?? "-")
            }
        }
        .padding(8)
        .background(Color(.tertiarySystemBackground))
        .cornerRadius(6)
    }

    private func tuningString(_ tuning: PitchPoint.Tuning) -> String {
        switch tuning {
        case .silent: return "SILENT"
        case .flat: return "FLAT"
        case .inTune: return "IN_TUNE"
        case .sharp: return "SHARP"
        default: return "UNKNOWN"
        }
    }

    private var apiInfoSection: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("PitchPoint Properties:")
                .font(.caption)
                .fontWeight(.medium)

            Text("• pitch: Float - Detected pitch in Hz (-1 if not singing)")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("• isSinging: Bool - True if voiced audio detected")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("• note: String? - Musical note name (e.g., \"A4\", \"C#5\")")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("• midiNote: Int - MIDI note number (e.g., 69 for A4)")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("• centsOff: Int - Deviation from note (-50 to +50)")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("• tuning: Tuning - SILENT, FLAT, IN_TUNE, or SHARP")
                .font(.caption2)
                .foregroundColor(.secondary)

            Text("• reliability: Float - Detection confidence (0.0-1.0)")
                .font(.caption2)
                .foregroundColor(.secondary)
        }
        .padding(8)
        .background(Color(.tertiarySystemBackground))
        .cornerRadius(6)
    }

    // MARK: - Audio Logic

    private func setupAudioIfNeeded() {
        guard detector == nil else { return }

        // Configure audio session for recording
        AudioSessionManager.configure(.recording)

        let algorithm = algorithms[selectedAlgorithm].algorithm
        let detectorConfig = PitchDetectorConfig.Builder()
            .algorithm(algorithm)
            .build()

        // SwiftF0 requires model provider
        var modelProvider: (() -> KotlinByteArray)? = nil
        if algorithm == .swiftF0 {
            if !modelLoaderConfigured {
                ModelLoader.configure()
                modelLoaderConfigured = true
            }
            modelProvider = { ModelLoader.loadSwiftF0() }
        }

        detector = CalibraPitch.createDetector(config: detectorConfig, modelProvider: modelProvider)

        let tempPath = FileManager.default.temporaryDirectory
            .appendingPathComponent("pitch_explorer_temp.m4a").path
        recorder = SonixRecorder.create(outputPath: tempPath, config: .voice)
    }

    private func cleanup() {
        recorder?.stop()
        recorder?.release()
        recorder = nil

        detector?.release()
        detector = nil
    }

    private func startRecording() {
        setupAudioIfNeeded()
        guard let recorder = recorder, let detector = detector else { return }

        detector.reset()
        isRecording = true

        Task {
            let hwRate = AudioSessionManager.hardwareSampleRate

            for await buffer in recorder.audioBuffers {
                let samples16k = SonixResampler.resample(
                    samples: buffer.samples,
                    fromRate: hwRate,
                    toRate: 16000
                )

                let result = detector.detect(buffer: samples16k)
                let amp = detector.getAmplitude(buffer: samples16k)

                await MainActor.run {
                    self.currentPitch = result
                    self.amplitude = amp
                }
            }
        }

        recorder.start()
    }

    private func stopRecording() {
        recorder?.stop()
        isRecording = false

        detector?.reset()
        currentPitch = nil
        amplitude = 0.0
    }
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/calibra/pitch/PitchPointExplorerDemo.kt
package com.musicmuni.voxatrace.demo.calibra.pitch

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.musicmuni.voxatrace.calibra.CalibraPitch
import com.musicmuni.voxatrace.calibra.model.PitchAlgorithm
import com.musicmuni.voxatrace.calibra.model.PitchPoint
import com.musicmuni.voxatrace.calibra.model.PitchPreset
import com.musicmuni.voxatrace.sonix.AudioSessionManager
import com.musicmuni.voxatrace.sonix.SonixRecorder
import com.musicmuni.voxatrace.sonix.SonixResampler
import kotlinx.coroutines.launch

/**
 * PitchPoint properties inspector demo.
 *
 * Showcases all computed properties of PitchPoint in real-time:
 * - pitch: Detected frequency in Hz
 * - isSinging: Voice activity detection
 * - note: Musical note name (A4, C#5, etc.)
 * - midiNote: MIDI note number
 * - centsOff: Deviation from nearest note
 * - tuning: SILENT, FLAT, IN_TUNE, SHARP
 * - reliability: Detection confidence
 */
@Composable
fun PitchPointExplorerDemo() {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    var detector by remember { mutableStateOf<CalibraPitch.Detector?>(null) }
    var recorder by remember { mutableStateOf<SonixRecorder?>(null) }
    var isRecording by remember { mutableStateOf(false) }
    var currentPitch by remember { mutableStateOf<PitchPoint?>(null) }
    var amplitude by remember { mutableFloatStateOf(0f) }

    fun startRecording() {
        scope.launch {
            val recordPath = "${context.cacheDir}/pitch_explorer.m4a"
            recorder?.release()
            recorder = SonixRecorder.create(recordPath, "m4a", "voice")

            detector?.release()
            detector = CalibraPitch.DetectorBuilder()
                .algorithm(PitchAlgorithm.YIN)
                .preset(PitchPreset.BALANCED)
                .build()

            recorder?.start()
            isRecording = true

            launch {
                recorder?.audioBuffers?.collect { buffer ->
                    val det = detector ?: return@collect

                    val hwRate = AudioSessionManager.hardwareSampleRate.toInt()
                    val samples16k = SonixResampler.resample(
                        samples = buffer.data,
                        fromRate = hwRate,
                        toRate = 16000
                    )

                    currentPitch = det.detect(samples16k)
                    amplitude = det.getAmplitude(samples16k)
                }
            }
        }
    }

    fun stopRecording() {
        scope.launch {
            recorder?.stop()
            recorder?.release()
            recorder = null
            isRecording = false
            currentPitch = null
            amplitude = 0f
        }
    }

    DisposableEffect(Unit) {
        onDispose {
            recorder?.release()
            detector?.release()
        }
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text(
            text = "PitchPoint Properties",
            style = MaterialTheme.typography.titleMedium
        )

        Text(
            text = "See all computed properties of PitchPoint in real-time",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        // Large note display
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.surfaceVariant
            )
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(24.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                if (currentPitch != null && currentPitch!!.isSinging) {
                    Text(
                        text = currentPitch!!.note ?: "-",
                        fontSize = 60.sp,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.primary
                    )
                    Text(
                        text = "%.1f Hz".format(currentPitch!!.pitch),
                        style = MaterialTheme.typography.titleLarge,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )

                    Spacer(modifier = Modifier.height(12.dp))

                    // Tuning indicator bar
                    TuningBar(
                        centsOff = currentPitch!!.centsOff,
                        tuning = currentPitch!!.tuning
                    )
                } else {
                    Text(
                        text = "-",
                        fontSize = 60.sp,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.outline
                    )
                    Text(
                        text = "Not singing",
                        style = MaterialTheme.typography.titleLarge,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
        }

        HorizontalDivider()

        // Properties grid
        Text(
            text = "Properties Inspector",
            style = MaterialTheme.typography.titleSmall,
            fontWeight = FontWeight.Medium
        )

        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.surface
            )
        ) {
            Column(
                modifier = Modifier.padding(12.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                PropertyRow("pitch", currentPitch?.let { "%.1f Hz".format(it.pitch) } ?: "-")
                PropertyRow("isSinging", currentPitch?.isSinging?.toString() ?: "-")
                PropertyRow("note", currentPitch?.note ?: "-")
                PropertyRow("midiNote", currentPitch?.let { if (it.isSinging) it.midiNote.toString() else "-" } ?: "-")
                PropertyRow("centsOff", currentPitch?.let { if (it.isSinging) "${it.centsOff}" else "-" } ?: "-")
                PropertyRow("tuning", currentPitch?.let { tuningString(it.tuning) } ?: "-")
                PropertyRow("reliability", currentPitch?.let { "%.2f".format(it.reliability) } ?: "-")
                PropertyRow("confidence", currentPitch?.let { "%.2f".format(it.confidence) } ?: "-")
            }
        }

        HorizontalDivider()

        // Control button
        Button(
            onClick = { if (isRecording) stopRecording() else startRecording() },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text(if (isRecording) "Stop" else "Start")
        }

        // API documentation
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.tertiaryContainer.copy(alpha = 0.3f)
            )
        ) {
            Column(
                modifier = Modifier.padding(12.dp),
                verticalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                Text(
                    text = "PitchPoint Properties:",
                    style = MaterialTheme.typography.labelMedium,
                    fontWeight = FontWeight.Medium
                )
                Text("• pitch: Float - Detected pitch in Hz (-1 if silent)", style = MaterialTheme.typography.bodySmall)
                Text("• isSinging: Boolean - True if voiced audio", style = MaterialTheme.typography.bodySmall)
                Text("• note: String? - Musical note (A4, C#5)", style = MaterialTheme.typography.bodySmall)
                Text("• midiNote: Int - MIDI number (69 = A4)", style = MaterialTheme.typography.bodySmall)
                Text("• centsOff: Int - Deviation (-50 to +50)", style = MaterialTheme.typography.bodySmall)
                Text("• tuning: Tuning - SILENT/FLAT/IN_TUNE/SHARP", style = MaterialTheme.typography.bodySmall)
                Text("• reliability: Float - Detection confidence", style = MaterialTheme.typography.bodySmall)
            }
        }
    }
}

@Composable
private fun PropertyRow(name: String, value: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Text(
            text = name,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
        Text(
            text = value,
            style = MaterialTheme.typography.bodySmall,
            fontWeight = FontWeight.Medium
        )
    }
}

@Composable
private fun TuningBar(centsOff: Int, tuning: PitchPoint.Tuning) {
    val color = when (tuning) {
        PitchPoint.Tuning.IN_TUNE -> Color.Green
        PitchPoint.Tuning.FLAT, PitchPoint.Tuning.SHARP -> Color(0xFFFF9800)
        else -> Color.Gray
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        // Progress bar representation
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(24.dp)
        ) {
            LinearProgressIndicator(
                progress = { ((centsOff + 50) / 100f).coerceIn(0f, 1f) },
                modifier = Modifier
                    .fillMaxWidth()
                    .height(8.dp)
                    .align(Alignment.Center),
                color = color
            )
        }

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text("-50¢", style = MaterialTheme.typography.labelSmall)
            Text("0", style = MaterialTheme.typography.labelSmall, color = Color.Green)
            Text("+50¢", style = MaterialTheme.typography.labelSmall)
        }

        Spacer(modifier = Modifier.height(4.dp))

        Text(
            text = tuningString(tuning),
            style = MaterialTheme.typography.labelMedium,
            fontWeight = FontWeight.Medium,
            color = color
        )
    }
}

private fun tuningString(tuning: PitchPoint.Tuning): String {
    return when (tuning) {
        PitchPoint.Tuning.SILENT -> "SILENT"
        PitchPoint.Tuning.FLAT -> "FLAT"
        PitchPoint.Tuning.IN_TUNE -> "IN TUNE"
        PitchPoint.Tuning.SHARP -> "SHARP"
    }
}
